### DISCUSSION TITLE
Can't access to a private repository using VCPKG registry

### DISCUSSION THREAD

**Discussion by maria-korol** [microsoft/vcpkg#29951] [Q&A]
I've created a VCPKG custom registry from a private repository, which is using LFS for test files. To access the repo we're using JFROG credentials. 

I've discovered from this topic [#17272](https://github.com/microsoft/vcpkg/discussions/17272) that I should use **vcpkg_from_git** to access the private repo using git auth.

Here is my func:

```
vcpkg_from_git(    
    OUT_SOURCE_PATH <SOURCE_PATH>    
    URL https://github.com/organization-name/repository  
    REF c2151d8df3f3af46d9a01301685bc756c5830844
    HEAD_REF main
    LFS https://ha-us.dso.organization-name.net/artifactory/api/lfs/tool-gitlfs-dev 
)
```

After running
`vcpkg install packageName --overlay-ports=ports/packageName`

The installation of port s successful, as VCPKG is starting to fetch the private repo, but fails:

```
Command failed: "C:/Program Files/Git/mingw64/bin/git.exe" lfs fetch https://ha-us.dso.organization-name.net/artifactory/api/lfs/tool-gitlfs-dev c2151d8df3f3af46d9a01301685bc756c5830844
    Working Directory: C:/vcpkg/downloads/git-tmp
    Error code: 2
```

So to have more info and to check LFS, I manually run this lfs fetch command with GIT_TRACE=1 GIT_TRANSFER_TRACE=1 GIT_CURL_VERBOSE=1 attributes:
```
GIT_TRACE=1 GIT_TRANSFER_TRACE=1 GIT_CURL_VERBOSE=1 lfs fetch https://ha-us.dso.organization-name.net/artifactory/api/lfs/tool-gitlfs-dev c2151d8df3f3af46d9a01301685bc756c5830844
```
 in **C:/vcpkg/downloads/git-tmp** and in local clone of this private repository

In the second case, there was the next request:

`14:26:26.679761 trace git-lfs: HTTP: POST https://ha-us.dso.organization-name.net/artifactory/api/lfs/tool-gitlfs-dev/objects/batch
`
which returns firstly 
```
14:37:19.492059 trace git-lfs: HTTP: {
  "errors" : [ {
    "status" : 401,
    "message" : "Authorization Required"
  } ]
}
```
but then status 200 and objects were successfully downloaded.

But when I run the same command in **C:/vcpkg/downloads/git-tmp**, the request is
`14:26:26.679761 trace git-lfs: HTTP: POST https://ha-us.dso.organization-name.net/artifactory/api/lfs/tool-gitlfs-dev.git/info/lfs/objects/batch`

which returns
```
{
  "errors" : [ {
    "status" : 404,
    "message" : "Not Found"
  } ]
}
batch response: Repository or object not found: https://ha-us.dso.organization-name.net/artifactory/api/lfs/tool-gitlfs-dev.git/info/lfs/objects/batch
Check that it exists and that you have proper access to it
```

So the **question** is why different links are used in different POST requests, as vcpkg couldn't find objects with the second one.

In addition, shouldn't we use the **.lfsconfig** file instead of passing the link in vcpkg_from_git?
I mean we're fetching the repository. We should take .lfsconfig file and then use the link from it.

And also, is there any possibility to download objects not using lfs fetch?
When i don't provide LFS link in vpckg_from_git, it is still fetching the repo. GIT_LFS_SKIP_SMUDGE=1 is not helping either.

**Comment by luncliff** [comment]
Hi, If the host is a GitHub (or GitHub Enterprise) instance, would you consider using `AUTHORIZATION_TOKEN` with `vcpkg_from_github`?

https://github.com/microsoft/vcpkg/blob/c6d6efed3e9b4242765bfe1b5c5befffd85f7b92/scripts/cmake/vcpkg_from_github.cmake#L33-L36

For example,

```cmake
vcpkg_from_github(
    OUT_SOURCE_PATH SOURCE_PATH
    REPO ${org}/${repo}
    REF ${version}
    SHA512 0
    GITHUB_HOST "https://private.github-enterprise.com"
    AUTHORIZATION_TOKEN "ghp_1234abcd..." # personal access token with repo read access
)
```

And for vcpkg-registry works, I'm embedding token for vcpkg-configuration.json in CI build

```json
{
  "default-registry": null,
  "registries": [
    {
      "packages": [
        "zlib-ng"
      ],
      "kind": "git",
      "repository": "https://ghp_abcd1234@private.github-enterprise.com/my-team/vcpkg-registry",
      "baseline": "000000"
    }
  ]
}
```
